byzcoin/contracts/contract_deferred_test.go

```diff
@@ -0,0 +1,1938 @@
+					},
+					{
+						Name:  "index",
+						Value: indexBuf,
+					},
+				},
+			},
+			SignerCounter: []uint64{4},
+		}},
+	}
+	require.Nil(t, ctx.FillSignersAndSignWith(signer))
+
+	_, err = cl.AddTransaction(ctx)
+	require.Nil(t, err)
+
+	proposedTransaction.Instructions[1].SignerIdentities = append(proposedTransaction.Instructions[1].SignerIdentities, identity)
+	proposedTransaction.Instructions[1].Signatures = append(proposedTransaction.Instructions[1].Signatures, signature)
+	result.ProposedTransaction = proposedTransaction
+	resultBuf, err = protobuf.Encode(&result)
+	require.Nil(t, err)
+
+	pr, err = cl.WaitProof(byzcoin.NewInstanceID(myID.Slice()), 2*genesisMsg.BlockInterval, resultBuf)
+	require.Nil(t, err)
+	require.True(t, pr.InclusionProof.Match(myID.Slice()))
+
+	dataBuf, _, _, err = pr.Get(myID.Slice())
+	require.Nil(t, err)
+
+	result = byzcoin.DeferredData{}
+	err = protobuf.Decode(dataBuf, &result)
+	require.Nil(t, err)
+
+	// We can not do this test because the identities have to be compared using
+	// the Equal() method.
+	//require.Equal(t, result.ProposedTransaction, proposedTransaction)
+	require.Equal(t, len(result.ProposedTransaction.Instructions), 2)
+	require.Equal(t, result.ExpireBlockIndex, expireBlockIndexInt)
+	require.NotEmpty(t, result.ProposedTransaction.Instructions[1].SignerIdentities)
+	require.Equal(t, len(result.ProposedTransaction.Instructions[1].SignerIdentities), 1)
+	// This test won't work. But by using Equal() will.
+	// require.Equal(t, result.ProposedTransaction.Instructions[0].SignerIdentities[0], identity)
+	require.True(t, identity.Equal(&result.ProposedTransaction.Instructions[1].SignerIdentities[0]))
+
+	require.NotEmpty(t, result.ProposedTransaction.Instructions[1].Signatures)
+	require.Equal(t, len(result.ProposedTransaction.Instructions[1].Signatures), 1)
+	require.Equal(t, result.ProposedTransaction.Instructions[1].Signatures[0], signature)
+	// Default NumExecution should be 1
+	require.Equal(t, result.NumExecution, uint64(1))
+
+	require.NotEmpty(t, result.InstructionHashes)
+
+	local.WaitDone(genesisMsg.BlockInterval)
+
+	// ------------------------------------------------------------------------
+	// 4. Invoke an "execRoot" command
+	// ------------------------------------------------------------------------
+
+	ctx = byzcoin.ClientTransaction{
+		Instructions: []byzcoin.Instruction{{
+			InstanceID: myID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: byzcoin.ContractDeferredID,
+				Command:    "execProposedTx",
+			},
+			SignerCounter: []uint64{5},
+		}},
+	}
+	require.Nil(t, ctx.FillSignersAndSignWith(signer))
+
+	_, err = cl.AddTransaction(ctx)
+	require.Nil(t, err)
+
+	// Need to sleep because we can't predict the output (hence the 'nil')
+	time.Sleep(2 * genesisMsg.BlockInterval)
+	pr, err = cl.WaitProof(byzcoin.NewInstanceID(ctx.Instructions[0].DeriveID("").Slice()), 2*genesisMsg.BlockInterval, nil)
+	require.Error(t, err)
+
+	local.WaitDone(genesisMsg.BlockInterval)
+
+}
```

Comment:
Correct me if I'm wrong but... I expected to find a test where the proposed instruction depended on having a darc with an AND in it evaluate to true. The test would spawn the deferred, add one proof, try to exec, confirm it failed, add another proof that is not useful to satisfying the DARC, try exec, confirm failed, add the last signature needed to satisfy the darc, and then exec and confirm the proposed txn goes in.

Also, I didn't see a test for deletion?

This test can be added in your next commit, though.

Reply:
That's right. I'm keeping it on the todo list !